feat(migrations): Add migrations for CDC tables

lynnagara opened the pull request:
Adds the migrations for the grouped message and group assignees tables.

This PR depends on https://github.com/getsentry/snuba/pull/1185 as we need the ability to create replicated, unsharded local tables on a sharded cluster.

lynnagara committed:
feat(migrations): Add migrations for CDC tables

Adds the migrations for the grouped message and group assignees tables.

fpacifici commented on snuba/migrations/snuba_migrations/events/0005_groupedmessages.py:
```diff
@@ -0,0 +1,74 @@
+
+
+columns = [
+    # columns to maintain the dataset
```
Not sure this line in this comment is relevant anymore.

fpacifici commented on snuba/migrations/snuba_migrations/events/0005_groupedmessages.py:
```diff
@@ -0,0 +1,74 @@
+                storage_set=StorageSetKey.EVENTS,
+                table_name="groupedmessage_dist",
+                columns=columns,
+                engine=table_engines.Distributed(
+                    local_table_name="groupedmessage_local", sharding_key=None,
+                ),
```
This works if the cluster you create this table onto has one shard and many replicas. What happens if we run this statement (without sharding_key) on a cluster that has multiple shards ? 
Would that still work, break or just create a random shard (which may not work if we write twice the same record, which is possible)

lynnagara replied:
Do you know if we have this on a distributed cluster today, or just the local ones? If not, this should probably be omitted completely.

fpacifici replied:
Yes we have it on a cluster that has one shard and several replicas.

lynnagara replied:
Confirmed with James that we do want to keep this on the same cluster as the regular events one. The thinking was that: It will work, but the drawback is that it will multiply the write pressure. But we are not writing to it right now anyway.

fpacifici commented on snuba/migrations/snuba_migrations/events/0006_groupassignees.py:
```diff
@@ -0,0 +1,66 @@
+
+
+columns = [
+    # columns to maintain the dataset
```
same comment I had above

fpacifici reviewed (commented):
I have one concern on how we are going to deploy this on a distributed cluster. These tables should not be sharded so we should probably enforced somehow that the cluster has one shard only to be compatible with them.

lynnagara committed:
feat(migrations): Add unsharded option to MergeTree tables

Add an unsharded option for merge tree tables that replicates the same
table across all shards of a cluster instead of actually sharding the
table. This will be needed for the grouped message and group assignee
tables.

Updates the path to use the storage set name instead of requiring the
user to provide a "layer" macro. Also updates the format of the path
to make it more readable in Zookeeper.

This is not quite the format we have in prod on our existing tables.
Since our previously defined paths cannot be changed, we will need to
add some escape hatch where we can override these with our actual path
values later.

lynnagara committed:
Merge branch 'add-unsharded-option' into migrate-cdc-tables

lynnagara committed:
fix merge issues

lynnagara committed:
remove irrelevant comment

lynnagara committed:
test all storages

lynnagara committed:
Merge remote-tracking branch 'origin/master' into migrate-cdc-tables

lynnagara committed:
rename

lynnagara committed:
group message and group assignee should be unsharded

lynnagara committed:
Merge branch 'master' into migrate-cdc-tables

hydra-zim commented:
hi, there may be some compatibility issues when use ReplicatedMergeTree with same replica_path. 

![image](https://user-images.githubusercontent.com/13027048/128651723-04372240-d304-460d-ba0c-f9acbc322d4c.png)
